airbyte-integrations/connectors/destination-kafka/src/main/java/io/airbyte/integrations/destination/kafka/KafkaRecordConsumer.java

```diff
@@ -0,0 +1,156 @@
+                             ConfiguredAirbyteCatalog catalog,
+                             Consumer<AirbyteMessage> outputRecordCollector,
+                             NamingConventionTransformer nameTransformer) {
+    this.topicPattern = kafkaDestinationConfig.getTopicPattern();
+    this.topicMap = new HashMap<>();
+    this.producer = kafkaDestinationConfig.getProducer();
+    this.sync = kafkaDestinationConfig.isSync();
+    this.transactionalProducer = kafkaDestinationConfig.isTransactionalProducer();
+    this.catalog = catalog;
+    this.callback = (metadata, exception) -> {
+      if (exception != null) {
+        // TODO improve error management
+        LOGGER.error("Error sending message to topic", exception);
+      }
+    };
+    this.outputRecordCollector = outputRecordCollector;
+    this.nameTransformer = nameTransformer;
+  }
+
+  @Override
+  protected void startTracked() {
+    Map<AirbyteStreamNameNamespacePair, String> mapped = catalog.getStreams().stream()
+        .map(stream -> AirbyteStreamNameNamespacePair.fromAirbyteSteam(stream.getStream()))
+        .collect(Collectors.toMap(Function.identity(),
+            pair -> nameTransformer.getIdentifier(topicPattern
+                .replaceAll("\\{namespace}", Optional.ofNullable(pair.getNamespace()).orElse(""))
+                .replaceAll("\\{stream}", Optional.ofNullable(pair.getName()).orElse("")))));
+    topicMap.putAll(mapped);
+
+    if (transactionalProducer) {
+      producer.initTransactions();
+    }
+  }
+
+  @Override
+  protected void acceptTracked(AirbyteMessage airbyteMessage) throws Exception {
+    if (airbyteMessage.getType() == AirbyteMessage.Type.STATE) {
+      lastStateMessage = airbyteMessage;
+    } else if (airbyteMessage.getType() == AirbyteMessage.Type.RECORD) {
+      final AirbyteRecordMessage recordMessage = airbyteMessage.getRecord();
+
+      final String topic = topicMap.get(AirbyteStreamNameNamespacePair.fromRecordMessage(recordMessage));
+      final String key = UUID.randomUUID().toString();
+      final JsonNode value = Jsons.jsonNode(ImmutableMap.of(
+          JavaBaseConstants.COLUMN_NAME_AB_ID, key,
+          JavaBaseConstants.COLUMN_NAME_EMITTED_AT, recordMessage.getEmittedAt(),
+          JavaBaseConstants.COLUMN_NAME_DATA, recordMessage.getData()));
+      final ProducerRecord<String, JsonNode> record = new ProducerRecord<>(topic, key, value);
+
+      if (transactionalProducer) {
+        sendRecordInTransaction(record);
+      } else {
+        sendRecord(record);
+      }
+    } else {
+      LOGGER.warn("Unexpected message: " + airbyteMessage.getType());
+    }
+  }
+
+  private void sendRecord(ProducerRecord<String, JsonNode> record) throws Exception {
+    Future<RecordMetadata> future = producer.send(record, callback);
+    if (sync) {
+      future.get();
+    }
+  }
+
+  private void sendRecordInTransaction(ProducerRecord<String, JsonNode> record) throws Exception {
+    try {
+      producer.beginTransaction();
```

Comment:
what is the value of creating a transaction on a single record? isn't that equivalent to having no transactionality?

Reply:
Just removed this. Having ``acks=all`` and ``sync=true`` it's enough